packages/rest/src/validate.ts

```diff
@@ -1,9 +1,53 @@
-import { Program } from "@cadl-lang/compiler";
+import {
+  getKeyName,
+  getServiceNamespace,
+  Model,
+  navigateTypesInNamespace,
+  Program,
+} from "@cadl-lang/compiler";
 import { getAllHttpServices } from "./http/operations.js";
+import { reportDiagnostic } from "./lib.js";
+import { getParentResource, getResourceTypeKey } from "./resource.js";
+
+function checkForDuplicateResourceKeyNames(program: Program, model: Model) {
+  // If the model type is defined under the service namespace, check to
+  // that the parent resource type(s) don't have the same key name as the
+  // current resource type.
+  const keyNameSet = new Set<string>();
```

Comment:
sorry for not commenting on that before, I wrote a `DuplicateTracker` for discriminator the other day (its in discriminator-utils.ts so might need to move it to a common file) but it does take care of tracking all duplicates and make it easy to show the error on all the duplicate instances vs here I think it would only show up on the first duplicate

Reply:
`DuplicateTracker` looks pretty useful for this.  Right now the diagnostic only shows up on the parent property that the child key conflicts with.  Would certainly be nicer to write it out for all dupes, though.

If I were to move `DuplicateTracker`, should I do that in a separate PR or try to get it done in this one?

Reply:
Think this is fine in this PR but whatever you think works better

Reply:
Moved `DuplicateTracker` and used it to improve diagnostics for duplicate keys.